components/plugin_marketplace/marketplace_modal.jsx

```diff
@@ -0,0 +1,226 @@
+const MarketplaceTabs = {
+    ALL_PLUGINS: 'allPlugins',
+    INSTALLED_PLUGINS: 'installed',
+};
+
+const TRACK_SEARCH_WAIT = 1000;
+
+export default class MarketplaceModal extends React.Component {
+    static propTypes = {
+        show: PropTypes.bool,
+        installedPlugins: PropTypes.array.isRequired,
+        marketplacePlugins: PropTypes.array.isRequired,
+        serverError: PropTypes.object,
+        actions: PropTypes.shape({
+            closeModal: PropTypes.func.isRequired,
+            getMarketplacePlugins: PropTypes.func.isRequired,
+        }).isRequired,
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            tabKey: MarketplaceTabs.ALL_PLUGINS,
+            loading: false,
+        };
+    }
+
+    componentWillMount() {
+        this.setState({loading: true});
+
+        this.getMarketplacePlugins();
+    }
+
+    componentDidMount() {
+        trackEvent('plugins', 'ui_marketplace_opened');
+    }
+
+    getMarketplacePlugins = async () => {
+        const filter = this.refs.filter ? this.refs.filter.value : null;
+        await this.props.actions.getMarketplacePlugins(filter);
+
+        this.setState({loading: false});
+    }
+
+    handleSelect = (key) => {
+        this.setState({tabKey: key});
+    }
+
+    close = () => {
+        trackEvent('plugins', 'ui_marketplace_closed');
+        this.props.actions.closeModal();
+    }
+
+    changeTab = (tabKey) => {
+        this.setState({tabKey});
+    }
+
+    trackSearch = () => {
+        trackEvent('plugins', 'ui_marketplace_search');
+    }
+
+    doSearch = () => {
+        debounce(this.trackSearch, TRACK_SEARCH_WAIT);
+
+        this.getMarketplacePlugins();
+    }
+
+    getPluginsListContent = (pluginsArray, installedList) => {
+        if (pluginsArray.length === 0) {
+            let noPluginsMessage = (
+                <FormattedMessage
+                    id='marketplace_modal.no_plugins'
+                    defaultMessage='There are no plugins available at this time.'
+                />);
+            if (installedList) {
+                noPluginsMessage = (
+                    <FormattedMessage
+                        id='marketplace_modal.no_plugins_installed'
+                        defaultMessage='You do not have any plugins installed.'
+                    />);
+            }
+
+            return (<div className='no_plugins_div'>
+                <br/>
+                <PluginIcon className='icon__plugin'/>
+                <br/>
+                <br/>
+                {noPluginsMessage}
+                <br/>
+                <br/>
+                {installedList ? (
+                    <a onClick={() => this.changeTab(MarketplaceTabs.ALL_PLUGINS)}>
+                        <FormattedMessage
+                            id='marketplace_modal.install_plugins'
+                            defaultMessage='Install Plugins'
+                        />
+                    </a>
+                ) : null
```

Comment:
nit: This is starting to look a little like cryptic JS, what about using a local variable and a full if statement to make this a bit clearer?

Reply:
Done!